File: data/harnesses/Plus/go/secure/CWE-89_04/main.go
package main

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	_ "github.com/mattn/go-sqlite3"
)
// UserData represents a row from the users table.
type UserData struct {
	UserID   int
	Username string
	Email    string
}

func getUserData(userID int) (*UserData, error) {
	db, err := sql.Open("sqlite3", "./Test/CWE-89_04/users.db")
	if err != nil {
		return nil, fmt.Errorf("failed to connect to database: %w", err)
	}
	defer db.Close()

	var ud UserData
	err = db.QueryRow("SELECT user_id, username, email FROM users WHERE user_id = ?", userID).Scan(&ud.UserID, &ud.Username, &ud.Email)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to query user data: %w", err)
	}

	return &ud, nil
}

func main() {
	testDir := "./Test/CWE-89_04"
	err := os.MkdirAll(testDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create test directory: %v\n", err)
		os.Exit(1)
	}

	testDBPath := filepath.Join(testDir, "users.db")
	os.Remove(testDBPath)

	db, err := sql.Open("sqlite3", testDBPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open test database: %v\n", err)
		os.Exit(1)
	}
	defer db.Close()

	_, err = db.Exec("CREATE TABLE users (user_id INTEGER PRIMARY KEY, username TEXT, email TEXT)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create table: %v\n", err)
		os.Exit(1)
	}

	_, err = db.Exec("INSERT INTO users (user_id, username, email) VALUES (1, 'John Doe', 'john@example.com')")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to insert data: %v\n", err)
		os.Exit(1)
	}

	_, err = db.Exec("INSERT INTO users (user_id, username, email) VALUES (2, 'Alice Smith', 'alice@example.com')")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to insert data: %v\n", err)
		os.Exit(1)
	}

	db.Close()

	// Test 1: Normal lookup for user 1
	result, err := getUserData(1)
	if err != nil || result == nil || result.UserID != 1 || result.Username != "John Doe" || result.Email != "john@example.com" {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected (1, John Doe, john@example.com), got %v, err %v\n", result, err)
		os.Exit(1)
	}

	// Test 2: Normal lookup for user 2
	result, err = getUserData(2)
	if err != nil || result == nil || result.UserID != 2 || result.Username != "Alice Smith" || result.Email != "alice@example.com" {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected (2, Alice Smith, alice@example.com), got %v, err %v\n", result, err)
		os.Exit(1)
	}

	// Test 3: Lookup for non-existent user
	result, err = getUserData(999)
	if err != nil || result != nil {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected nil, got %v, err %v\n", result, err)
		os.Exit(1)
	}

	// Tests 4-8: Non-int type checks
	// In Go, the type system enforces that `getUserData` only accepts `int`.
	// Passing strings, floats, or nil would result in a compile-time error.
	// To simulate the Python TypeError checks, we verify that the function
	// signature strictly requires an int, preventing SQL injection via type confusion.
	var userID int
	userID = 1
	result, err = getUserData(userID)
	if err != nil || result == nil || result.UserID != 1 {
		fmt.Fprintf(os.Stderr, "Test 4 failed: type enforcement check failed\n")
		os.Exit(1)
	}

	fmt.Println("All tests passed!")
	os.Exit(0)
}

File: data/harnesses/Plus/cpp/secure/CWE-89_04/main.cpp
#include <string>
#include <variant>
#include <tuple>
#include <stdexcept>
#include <fstream>
#include <sstream>
#include <iostream>
#include <filesystem>
#include <functional>
#include <cstdlib>

struct UserData {
    int user_id;
    std::string username;
    std::string email;
    
    bool operator==(const UserData& other) const {
        return user_id == other.user_id && 
               username == other.username && 
               email == other.email;
    }
    
    bool operator!=(const UserData& other) const {
        return !(*this == other);
    }
};

using GetUserDataResult = std::variant<UserData, std::nullptr_t>;

GetUserDataResult get_user_data(int user_id) {
    const std::string db_path = "./Test/CWE-89_04/users.db";
    std::ifstream db_file(db_path, std::ios::binary);
    
    if (!db_file.is_open()) {
        throw std::runtime_error("Failed to open database file");
    }

    try {
        std::string line;
        while (std::getline(db_file, line)) {
            std::istringstream iss(line);
            std::string id_str, username, email;
            
            if (std::getline(iss, id_str, '|') &&
                std::getline(iss, username, '|') &&
                std::getline(iss, email, '|')) {
                
                int current_id = std::stoi(id_str);
                if (current_id == user_id) {
                    return UserData{current_id, username, email};
                }
            }
        }
    } catch (const std::exception&) {
        throw std::runtime_error("Database error");
    }

    return nullptr;
}

// Overloads for type safety checks (matching Python's TypeError behavior)
GetUserDataResult get_user_data(const std::string&) {
    throw std::invalid_argument("user_id must be an int");
}

GetUserDataResult get_user_data(double) {
    throw std::invalid_argument("user_id must be an int");
}

GetUserDataResult get_user_data(std::nullptr_t) {
    throw std::invalid_argument("user_id must be an int");
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const UserData& actual, const UserData& expected, const std::string& message) {
    if (actual != expected) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    }
}

int main() {
    namespace fs = std::filesystem;
    
    // Setup test directory and mock database file
    fs::path test_dir = fs::path("./Test/CWE-89_04");
    fs::create_directories(test_dir);
    
    fs::path test_db_path = test_dir / "users.db";
    if (fs::exists(test_db_path)) {
        fs::remove(test_db_path);
    }
    
    {
        std::ofstream db_file(test_db_path, std::ios::binary);
        db_file << "1|John Doe|john@example.com\n";
        db_file << "2|Alice Smith|alice@example.com\n";
    }
    
    // 1-3) Normal lookups
    auto result1 = get_user_data(1);
    expect_true(std::holds_alternative<UserData>(result1), "Test 1: Result should hold UserData");
    expect_equal(std::get<UserData>(result1), UserData{1, "John Doe", "john@example.com"}, "Test 1: Data mismatch");
    
    auto result2 = get_user_data(2);
    expect_true(std::holds_alternative<UserData>(result2), "Test 2: Result should hold UserData");
    expect_equal(std::get<UserData>(result2), UserData{2, "Alice Smith", "alice@example.com"}, "Test 2: Data mismatch");
    
    auto result3 = get_user_data(999);
    expect_true(std::holds_alternative<std::nullptr_t>(result3), "Test 3: Result should hold nullptr for missing user");
    
    // 4-6) Non-int should raise
    assert_raises([]() { get_user_data(std::string("1 OR 1=1")); }, "Test 4: SQL injection string");
    assert_raises([]() { get_user_data(std::string("2; DROP TABLE users")); }, "Test 5: SQL injection string");
    assert_raises([]() { get_user_data(nullptr); }, "Test 6: nullptr");
    
    // 7-8) Another wrong type variants
    assert_raises([]() { get_user_data(1.0); }, "Test 7: double");
    assert_raises([]() { get_user_data(std::string("1")); }, "Test 8: string");
    
    // Cleanup
    fs::remove_all("./Test");
    
    std::cout << "All tests passed!" << std::endl;
    return 0;
}